twister: deal with UnicodeEncodeErrors of serial output on acrn_ehl_crb

enjiamai opened the pull request:
When twister to decode ACRN's serial output, in the boot stage before
ACRN and zephyr starting to have serial output, there will have be
messages that cause twister trigger an "UnicodeEncodeError: 'latin-1'
codec can't encode character '\ufffc' in position 3127: ordinal not
in range(256)". We try to catch this exception and dynamically ignore
it to prevent the error.

Fixes #42548.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

nashif commented on scripts/twister:
```diff
@@ -772,6 +772,12 @@ structure in the main Zephyr tree: boards/<arch>/<board_name>/""")
              "stdout detailing RAM/ROM sizes on the specified filenames. "
              "All other command line arguments ignored.")
 
+    parser.add_argument("--serial-decode-specified", action="store_true",
```
do we really need an option for this? Can you please implement it  generically so that it does not require an option? I am not sure why you tie this to a board, this is probably related to your environment where you flash and test this board, it has nothing to do with the board.

enjiamai replied:
Hi @nashif , it is related to acrn_ehl_crb serial output, other board such as ehl_crb or up2 does not have this problem. About the environment, I'll try again to see if we can have another way to avoid this. I just thought implementing an option is a more generic way than adding things like "if self.board == ''acrn_ehl_crb" in the code. Seems like my idea is not correct. Thanks for your review!

dcpleung replied:
Wouldn't it be easier to do try/except in this case? If utf-8 or default encoding doesn't work, try latin-1, or iso-8859-1.

enjiamai replied:
Hi @dcpleung , good suggestion, will change to this, thx!

hakehuang commented on scripts/twister:
```diff
@@ -772,6 +772,12 @@ structure in the main Zephyr tree: boards/<arch>/<board_name>/""")
              "stdout detailing RAM/ROM sizes on the specified filenames. "
              "All other command line arguments ignored.")
 
+    parser.add_argument("--serial-decode-specified", action="store_true",
```
@enjiamai , please add this option in alpha-beta order by the leding character, in this case it is 's'

enjiamai replied:
Hi @hakehuang , I think I won't use an extra args/options in the coming update. But you are right, it should follow this alpha-beta order if an option need to be added. Thanks for the reminder!

enjiamai committed:
twister: deal with UnicodeEncodeErrors of serial output on acrn_ehl_crb

When twister to decode ACRN's serial output, in the boot stage before
ACRN and zephyr starting to have serial output, there will have be
messages that cause twister trigger an "UnicodeEncodeError: 'latin-1'
codec can't encode character '\ufffc' in position 3127: ordinal not
in range(256)". We try to catch this exception and ignore it.

Fixes #42548.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

chen-png commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -681,16 +681,19 @@ def monitor_serial(self, ser, halt_fileno, harness):
                 sl = serial_line.decode('utf-8', 'ignore').lstrip()
-                logger.debug("DEVICE: {0}".format(sl.rstrip()))
+                try:
+                    logger.debug("DEVICE: {0}".format(sl.rstrip()))
```
will this report UnicodeEncodeError ? seems like all invalid chars are already handled by line above "sl = serial_line.decode('utf-8', 'ignore').lstrip()", the output sl is normal, it won't meet any encode/decode error.

enjiamai replied:
Hi, Peng, no, it doesn't work and still caught an UnicodeEncodeError exception on this. I think this message might comes from grub stage.

enjiamai replied:
And this only happened at that certain moment once, the rest of the message doesn't lead to such an error.

enjiamai commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -2217,7 +2221,11 @@ def run_cmake(self, args=[]):
         if out:
             with open(os.path.join(self.build_dir, self.log), "a") as log:
                 log_msg = out.decode(sys.getdefaultencoding())
```
I added it, but it doesn't work. Actually we did not meet this problem previous. This is in cmake build stage, and we got an exception here since we start to use logging subsystem v2. Any idea for this?

dcpleung commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -681,16 +681,19 @@ def monitor_serial(self, ser, halt_fileno, harness):
             # is available yet.
             if serial_line:
                 sl = serial_line.decode('utf-8', 'ignore').lstrip()
```
The `.decode()` needs to be inside the `try` block so exception can be caught.

enjiamai replied:
Hi @dcpleung, thanks for the review. Actually the exception happens when writing the log, not in decode phasing. If this is the case, do you think I still need to put this line in try block?

dcpleung replied:
if that's the case, changing the encoding when the file is opened would probably work.

enjiamai replied:
Ok, let me give it a try, thanks!

Another info FYI:  this only happens in ACRN, (actually it happens in grub, I guess grub output some unknow control character, and it cause problem when we are printing it or writing to file).

enjiamai replied:
Hi @dcpleung, do you think could I still apply this? I think get a exception and re-open a file with a new encode will need more code to do this. I want just catch exception and skip it because this is not a valid message, and only happened once.

dcpleung replied:
TBH, changing the decoding from `utf-8` to `latin-1` here would probably be better. Or open the file as `utf-8` instead. The error comes from mismatched encodings between the opened file, and the line being written to the file. So make them match and the issue should go away. (Note that you will still want to use try/except to catch any error there and print an error message).

enjiamai replied:
Thanks so much for your great suggestion, @dcpleung, I tried to implement the you mentioned, it works, like this:
```
diff --git a/scripts/pylib/twister/twisterlib.py b/scripts/pylib/twister/twisterlib.py
index af6ee87f43..474032d77e 100755
--- a/scripts/pylib/twister/twisterlib.py
+++ b/scripts/pylib/twister/twisterlib.py
@@ -645,7 +645,7 @@ class DeviceHandler(Handler):
             harness.handle(None)
             return

-        log_out_fp = open(self.log, "wt")
+        log_out_fp = open(self.log, encoding='utf-8', mode="wt")

         ser_fileno = ser.fileno()
         readlist = [halt_fileno, ser_fileno]
@@ -680,7 +680,11 @@ class DeviceHandler(Handler):
             # Just because ser_fileno has data doesn't mean an entire line
             # is available yet.
             if serial_line:
-                sl = serial_line.decode('utf-8', 'ignore').lstrip()
+                try:
+                    sl = serial_line.decode('utf-8', 'ignore').lstrip()
+                except UnicodeEncodeError:
+                    sl = serial_line.decode('latin-1', 'ignore').lstrip()
+
                 logger.debug("DEVICE: {0}".format(sl.rstrip()))

                 log_out_fp.write(sl)
@@ -2217,8 +2221,8 @@ class CMake():
             results = {"returncode": p.returncode}

         if out:
-            with open(os.path.join(self.build_dir, self.log), "a") as log:
-                log_msg = out.decode(sys.getdefaultencoding())
+            with open(os.path.join(self.build_dir, self.log), encoding='utf-8', mode="a") as log:
+                log_msg = out.decode(sys.getdefaultencoding(), 'ignore')
                 log.write(log_msg)

         return results
```
And during implement this, I was eventually aware of what the problem is. I appreciate the review from you and other reviewers. Sorry the waste your time.

dcpleung commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -2217,7 +2220,10 @@ def run_cmake(self, args=[]):
         if out:
             with open(os.path.join(self.build_dir, self.log), "a") as log:
                 log_msg = out.decode(sys.getdefaultencoding())
```
`.decode()` needs to be inside `try` block.

enjiamai replied:
This line is also as the same as line 683, the exception did not trigger here. Looking forward to your opinion, thanks!

hakehuang commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -681,16 +681,19 @@ def monitor_serial(self, ser, halt_fileno, harness):
+                    logger.debug("DEVICE: {0}".format(sl.rstrip()))
+                    log_out_fp.write(sl)
+                except UnicodeEncodeError:
+                    continue
```
should we just ignore those chars? continue means skip all belows
```suggestion
                    Pass
```

hakehuang reviewed (dismissed):
add some suggestions.

enjiamai commented:
Thanks for all the reviewers @dcpleung @hakehuang @chen-png and sorry for spending your time to review this.  Actually @nashif point out the problem at first already but I was not aware of that (environment problem). Just export the LANG=xxx.utf-8 solve the problem. So close this.
